### ISSUE TITLE
[subinterpreters] crash when importing _sre in subinterpreters in parallel (Python 3.9 regression)

### ISSUE THREAD

**Issue by 0ce8d42d-8f7b-425f-94e3-f5f6d3d827df** [python/cpython#90228]
BPO | [46070](https://bugs.python.org/issue46070)
--- | :---
Nosy | @vstinner, @encukou, @ericsnowcurrently, @ndjensen, @hroncok, @uckelman, @corona10, @miss-islington, @shihai1991, @erlend-aasland, @bsteffensmeier, @M-Reimer, @prahal
PRs | <li>python/cpython#30423</li><li>python/cpython#30453</li><li>python/cpython#30454</li><li>python/cpython#30564</li><li>python/cpython#30565</li><li>python/cpython#30566</li><li>python/cpython#30577</li><li>python/cpython#30578</li><li>python/cpython#30579</li><li>python/cpython#30580</li>
Files | <li>[win_py399_crash_reproducer.py](https://bugs.python.org/file50494/win_py399_crash_reproducer.py "Uploaded as text/plain at 2021-12-15.18:35:10 by @bsteffensmeier"): Reproducer that intermittently crashes on windows</li><li>[bug.py](https://bugs.python.org/file50497/bug.py "Uploaded as text/plain at 2021-12-16.00:40:46 by @vstinner")</li><li>[pyobject_ob_interp.patch](https://bugs.python.org/file50560/pyobject_ob_interp.patch "Uploaded as text/plain at 2022-01-13.20:24:18 by @vstinner")</li><li>[sqlite3_crash.py](https://bugs.python.org/file50699/sqlite3_crash.py "Uploaded as text/plain at 2022-03-24.16:27:15 by @prahal"): Reproducer that crashes 90% of the time</li><li>[bug.py_asyncio_cpustressed-crash.log](https://bugs.python.org/file50700/bug.py_asyncio_cpustressed-crash.log "Uploaded as text/x-log at 2022-03-24.21:16:20 by @prahal"): asyncio bug.py crash while running stress -c `nproc --all`</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-12-14.10:57:50.149>
labels = ['interpreter-core', 'type-bug', '3.9', '3.10', '3.11', 'expert-subinterpreters']
title = '[subinterpreters] crash when importing _sre in subinterpreters in parallel (Python 3.9 regression)'
updated_at = <Date 2022-03-24.21:16:20.771>
user = 'https://bugs.python.org/graysky'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-24.21:16:20.771>
actor = 'prahal'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core', 'Subinterpreters']
creation = <Date 2021-12-14.10:57:50.149>
creator = 'graysky'
dependencies = []
files = ['50494', '50497', '50560', '50699', '50700']
hgrepos = []
issue_num = 46070
keywords = ['patch']
message_count = 54.0
messages = ['408520', '408539', '408540', '408541', '408558', '408568', '408633', '408662', '408664', '408665', '409245', '409255', '409461', '409573', '409686', '409772', '409778', '409780', '409795', '409798', '409802', '409962', '409965', '409966', '409970', '409980', '409983', '409984', '409991', '410010', '410014', '410442', '410444', '410446', '410447', '410493', '410497', '410498', '410500', '410505', '410507', '410509', '410510', '410513', '410517', '410518', '410520', '415954', '415955', '415963', '415964', '415970', '415975', '415977']
nosy_count = 15.0
nosy_names = ['vstinner', 'petr.viktorin', 'eric.snow', 'ndjensen', 'hroncok', 'uckelman', 'corona10', 'miss-islington', 'shihai1991', 'erlendaasland', 'graysky', 'bsteffensmeier', 'M-Reimer', 'jokot3', 'prahal']
pr_nums = ['30423', '30453', '30454', '30564', '30565', '30566', '30577', '30578', '30579', '30580']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue46070'
versions = ['Python 3.9', 'Python 3.10', 'Python 3.11']
```

</p></details>

**Comment by vstinner** [comment]
So far, only released Python v3.11.0a4 include my "_PyGC_Fini() untracks objects" fix. The 3.9 and 3.10 branch include my fix, but there is no release including it yet.

Scheduled releases:

* Python 3.9.13: Monday, 2022-05-16: https://peps.python.org/pep-0596/
* Python 3.10.5: Monday, 2022-06-06: https://peps.python.org/pep-0619/

> I managed to reproduce the bug.py crash with main branch up to commit https://github.com/python/cpython/commit/12c0012cf97d21bc637056983ede0eaf4c0d9c33 .

Oh really? This version include my fix. I cannot reproduce it on the current main branch (commit ea2f5bcda1a392804487e6883be89fbad38a01a5) nor at commit 12c0012cf97d21bc637056983ede0eaf4c0d9c33. Which commands did you type to build Python and reproduce the crash? What is your OS and C compiler?

**Comment by vstinner** [comment]
The original issue is fixed. Please open a new issue for sqlite3_crash.py.

**Comment by vstinner** [comment]
I created https://github.com/python/cpython/issues/92036 for sqlite3_crash.py bug.

**Comment by erlend-aasland** [comment]
Should this be closed then?

**Comment by vstinner** [comment]
I kept this issue open for this reason:

> It would be nice to add some tests.

**Comment by vstinner** [comment]
My notes about this bug: https://twitter.com/VictorStinner/status/1481783100191428611

**Comment by serhiy-storchaka** [comment]
`_sre` was converted to multi-phase init with per-interpreter state, which removed the root cause. Importing `_sre` in 160 subinterpreters across 8 parallel threads runs cleanly on a debug build of main. Closing as fixed.
